File: servers/zig/src/db/postgres.zig
const std = @import("std");
const pg = @import("pg");
const uuid = @import("../uuid.zig");
const user = @import("../user.zig");

const User = user.User;

/// Postgres repository backed by pg.zig's native connection pool. Pool size
/// (50) matches the other servers in the suite for fairness.
pub const Postgres = struct {
    pool: *pg.Pool,

    pub fn init(io: std.Io, allocator: std.mem.Allocator, url: []const u8) !Postgres {
        const uri = try std.Uri.parse(url);
        const pool = try pg.Pool.initUri(io, allocator, uri, .{ .size = 50, .timeout = 10_000 });
        return .{ .pool = pool };
    }

    pub fn deinit(self: *Postgres) void {
        self.pool.deinit();
    }

    pub fn health(self: *Postgres) bool {
        _ = self.pool.exec("SELECT 1", .{}) catch return false;
        return true;
    }

    pub fn create(self: *Postgres, arena: std.mem.Allocator, data: user.CreateUser) !User {
        var id_buf: [36]u8 = undefined;
        uuid.v7(&id_buf);
        const id = try arena.dupe(u8, &id_buf);
        _ = try self.pool.exec(
            "INSERT INTO users (id, name, email, favorite_number) VALUES ($1::uuid, $2, $3, $4)",
            .{ id, data.name, data.email, data.favoriteNumber },
        );
        return .{ .id = id, .name = data.name, .email = data.email, .favoriteNumber = data.favoriteNumber };
    }

    pub fn find(self: *Postgres, arena: std.mem.Allocator, id: []const u8) !?User {
        if (!uuid.isValid(id)) return null;
        var row = (try self.pool.row(
            "SELECT id::text, name, email, favorite_number FROM users WHERE id = $1::uuid",
            .{id},
        )) orelse return null;
        defer row.deinit() catch {};
        return try rowToUser(arena, &row);
    }

    pub fn update(self: *Postgres, arena: std.mem.Allocator, id: []const u8, data: user.UpdateUser) !?User {
        if (!uuid.isValid(id)) return null;
        var row = (try self.pool.row(
            "UPDATE users SET name = COALESCE($2, name), email = COALESCE($3, email), " ++
                "favorite_number = COALESCE($4, favorite_number) WHERE id = $1::uuid " ++
                "RETURNING id::text, name, email, favorite_number",
            .{ id, data.name, data.email, data.favoriteNumber },
        )) orelse return null;
        defer row.deinit() catch {};
        return try rowToUser(arena, &row);
    }

    pub fn delete(self: *Postgres, id: []const u8) !bool {
        if (!uuid.isValid(id)) return false;
        const affected = (try self.pool.exec("DELETE FROM users WHERE id = $1::uuid", .{id})) orelse 0;
        return affected > 0;
    }

    pub fn deleteAll(self: *Postgres) !void {
        _ = try self.pool.exec("DELETE FROM users", .{});
    }

    fn rowToUser(arena: std.mem.Allocator, row: *pg.QueryRow) !User {
        return .{
            .id = try arena.dupe(u8, try row.get([]const u8, 0)),
            .name = try arena.dupe(u8, try row.get([]const u8, 1)),
            .email = try arena.dupe(u8, try row.get([]const u8, 2)),
            .favoriteNumber = try row.get(?i32, 3),
        };
    }
};

File: shared/rust/src/db/postgres.rs
//! Postgres repository (sqlx, runtime query API).
//!
//! We use sqlx's runtime `query_as`/`query` rather than the compile-time `query!`
//! macros so the crate builds with no live database and no committed `.sqlx`
//! offline cache — the contract container has no dev DB at image-build time. The
//! pool is pinned to `max_connections(50)` to match the Go/TS Postgres pools
//! (fairness canon).

use sqlx::postgres::{PgPool, PgPoolOptions};
use uuid::Uuid;

use crate::error::DbError;
use crate::model::{CreateUser, UpdateUser, User};

#[derive(sqlx::FromRow)]
struct UserRow {
    id: Uuid,
    name: String,
    email: String,
    favorite_number: Option<i32>,
}

impl From<UserRow> for User {
    fn from(row: UserRow) -> Self {
        Self {
            id: row.id.to_string(),
            name: row.name,
            email: row.email,
            favorite_number: row.favorite_number,
        }
    }
}

pub struct PostgresRepo {
    pool: PgPool,
}

impl PostgresRepo {
    pub async fn connect(url: &str) -> Result<Self, DbError> {
        let pool = PgPoolOptions::new()
            .max_connections(50)
            .connect(url)
            .await?;
        Ok(Self { pool })
    }

    pub async fn create(&self, data: &CreateUser) -> Result<User, DbError> {
        let id = Uuid::now_v7();
        let row = sqlx::query_as::<_, UserRow>(
            "INSERT INTO users (id, name, email, favorite_number) VALUES ($1, $2, $3, $4) \
             RETURNING id, name, email, favorite_number",
        )
        .bind(id)
        .bind(&data.name)
        .bind(&data.email)
        .bind(data.favorite_number)
        .fetch_one(&self.pool)
        .await?;
        Ok(row.into())
    }

    pub async fn find_by_id(&self, id: &str) -> Result<Option<User>, DbError> {
        // An unparseable id is simply "not found", never an error.
        let Ok(uuid) = Uuid::parse_str(id) else {
            return Ok(None);
        };
        let row = sqlx::query_as::<_, UserRow>(
            "SELECT id, name, email, favorite_number FROM users WHERE id = $1",
        )
        .bind(uuid)
        .fetch_optional(&self.pool)
        .await?;
        Ok(row.map(Into::into))
    }

    pub async fn update(&self, id: &str, data: &UpdateUser) -> Result<Option<User>, DbError> {
        if data.is_empty() {
            return self.find_by_id(id).await;
        }
        let Ok(uuid) = Uuid::parse_str(id) else {
            return Ok(None);
        };
        let row = sqlx::query_as::<_, UserRow>(
            "UPDATE users SET \
             name = COALESCE($2, name), \
             email = COALESCE($3, email), \
             favorite_number = COALESCE($4, favorite_number) \
             WHERE id = $1 \
             RETURNING id, name, email, favorite_number",
        )
        .bind(uuid)
        .bind(data.name.as_deref())
        .bind(data.email.as_deref())
        .bind(data.favorite_number)
        .fetch_optional(&self.pool)
        .await?;
        Ok(row.map(Into::into))
    }

    pub async fn delete(&self, id: &str) -> Result<bool, DbError> {
        let Ok(uuid) = Uuid::parse_str(id) else {
            return Ok(false);
        };
        let rows = sqlx::query("DELETE FROM users WHERE id = $1")
            .bind(uuid)
            .execute(&self.pool)
            .await?
            .rows_affected();
        Ok(rows > 0)
    }

    pub async fn delete_all(&self) -> Result<(), DbError> {
        sqlx::query("DELETE FROM users").execute(&self.pool).await?;
        Ok(())
    }

    pub async fn health_check(&self) -> Result<bool, DbError> {
        sqlx::query("SELECT 1").execute(&self.pool).await?;
        Ok(true)
    }

    pub async fn disconnect(&self) {
        self.pool.close().await;
    }
}
